# Recuperación Entregables Mayo 2025

Francisco Javier Mercader Martínez

## Machine Learning. Grado en Ciencia de datos

Los siguientes ejercicios tienen que ser entregados dentro de las dos horas de prácticas. Como realizar el entregable:


-   La realización se debe de hacer de forma  **INDIVIDUAL**
-   Se debe de enviar un notebook con el código y las explicaciones, comentarios, análisis y justificaciones en Markdown.
-   Reproducibilidad:
    -   Se debe de establecer una variable semilla con un número al inicio, esta variables será la que se utilice para el np.random_seed y para el random_state.
    -   Las bases de datos se cargan con rutas relativas.
-   El notebook se debe de subir a la tarea del aula virtual creada antes de la finalización de la hora de clase. (Ver tarea y fecha de cierre).

### DATASET

El conjunto de datos DIETA.csv describe los siguientes atributos:

- GENERO: Género de la persona. 0: Hombre, 1:Mujer
- EDAD: Edad en años
- ALTURA: ALtura de la persona en metros
- PESO: Peso de la persona en kg
- PRIMERACONSULTA: Si hay antecedentes de obesidad en la familia. 0:no, 1:si
- CBASURA: Si come comida basura. 0:no, 1:si
- VERDURA: Media de veces que come verdura durante la semana
- NCOMIDAS: Media de comidas que hace al día
- DEPORTE: Si hace ejecicio de fuerza: no, TodosDias, AlgunaVez, RaraVez
- FUMAS: Si el paciente fuma. 0:no, 1:si
- LAGUA: Media de Litros de agua que bebe al día
- CCALORIAS: Control de las calorias en las comidas. 0:no, 1:si
- EJERCICIO: Media de veces que hace ejercicio de cardio a la semana
- TECNOLOGIA: Media de días que usa tecnología a la semana
- DIETA: Si el paciente hace dieta o no. 1:no, 0:si

## Común para todos los ejercicios

1. Cargue el conjunto de datos.

## ENTREGABLE I

1. Se requiere realizar un modelo utilizando Random Forest y un árbol de decisión para predecir la variable **Dieta**. Sabiendo esto, ¿el modelo será de clasificación o de regresión? Analiza los datos y transformarlos si fuera necesario. Justifica las decisiones que tomes.

In [1]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder

# Cargar los datos
df = pd.read_csv('Dieta.csv', sep=';')

print("Datos originales:")
print(df.info())

# Transformar la variable 'DEPORTE' con OneHotEncoder
ohe = OneHotEncoder(sparse_output=False, drop='first')
deporte_encoded = ohe.fit_transform(df[['DEPORTE']])
derpote_cols = ohe.get_feature_names_out(['DEPORTE'])
df_deporte = pd.DataFrame(deporte_encoded, columns=derpote_cols, index=df.index)

# Concatenar y eliminar la columna original
df = pd.concat([df.drop('DEPORTE', axis=1), df_deporte], axis=1)

# Comprobar la transformación
print("\nDatos transformados:")
print(df.dtypes)

Datos originales:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1234 entries, 0 to 1233
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   GENERO           1234 non-null   int64  
 1   EDAD             1234 non-null   int64  
 2   ALTURA           1234 non-null   float64
 3   PESO             1234 non-null   float64
 4   PRIMERACONSULTA  1234 non-null   int64  
 5   CBASURA          1234 non-null   int64  
 6   VERDURA          1234 non-null   float64
 7   NCOMIDAS         1234 non-null   float64
 8   DEPORTE          1234 non-null   object 
 9   FUMAS            1234 non-null   int64  
 10  LAGUA            1234 non-null   float64
 11  CCALORIAS        1234 non-null   int64  
 12  EJERCICIO        1234 non-null   float64
 13  TECNOLOGIA       1234 non-null   float64
 14  DIETA            1234 non-null   int64  
dtypes: float64(7), int64(7), object(1)
memory usage: 144.7+ KB
None

Datos transformados:
GE

2. Realiza un hold-out de 70-30 y crea un modelo utilizando un árbol de decisión estableciendo como mínimo 2 parámetros. Utiliza dos métricas. Los resultados se deben de obtener en test y en train. ¿Hay sobreajuste o subajuste? Justifica la respuesta.

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, f1_score

# Separar los datos en variables predictoras y objetivo
X = df.drop('DIETA', axis=1)
y = df['DIETA']

# Hold-out 70-30
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

# Crear y entrenar el modelo
clf = DecisionTreeClassifier(max_depth=4, min_samples_split=10, random_state=42)
clf.fit(X_train, y_train)

# Predicciones
y_train_pred = clf.predict(X_train)
y_test_pred = clf.predict(X_test)

# Métricas
acc_train = accuracy_score(y_train, y_train_pred)
acc_test = accuracy_score(y_test, y_test_pred)
f1_train = f1_score(y_train, y_train_pred)
f1_test = f1_score(y_test, y_test_pred)

print(f"Accuracy:\n\t - Train: {acc_train:.4f} \n\t - Test: {acc_test:.4f}")
print(f"F1 Score:\n\t - Train: {f1_train:.4f}\n\t - Test: {f1_test:.4f}")

Accuracy:
	 - Train: 0.9954 
	 - Test: 0.9811
F1 Score:
	 - Train: 0.9958
	 - Test: 0.9827


Los resultados muestran un rendimiento muy alto tanto en `train` como en `test`.

Debido a que la diferencia entre ambos es mínima, **no hay sobreajuste** ni **subajeste**. Por lo tanto podemos afirmar que el modelo generaliza bien.

3. Realiza un ajuste de dos hiperparámetros para un modelo de Random Forest ¿Qué resultados obtienes, son mejores que los del árbol de la pregunta anterior? ¿Algún modelo es factible para saber si una persona necesita dieta o no?

In [3]:
from sklearn.ensemble import RandomForestClassifier

# Ajuste de hiperparámetros
rf = RandomForestClassifier(n_estimators=100, max_depth=6, random_state=42)
rf.fit(X_train, y_train)

# Predicciones
y_train_pred_rf = rf.predict(X_train)
y_test_pred_rf = rf.predict(X_test)

# Métricas
acc_train_rf = accuracy_score(y_train, y_train_pred_rf)
acc_test_rf = accuracy_score(y_test, y_test_pred_rf)
f1_train_rf = f1_score(y_train, y_train_pred_rf)
f1_test_rf = f1_score(y_test, y_test_pred_rf)

print(f"Random Forest Accuracy:\n\t - Train: {acc_train_rf:.4f} \n\t - Test: {acc_test_rf:.4f}")
print(f"Random Forest F1 Score:\n\t - Train: {f1_train_rf:.4f}\n\t - Test: {f1_test_rf:.4f}")


Random Forest Accuracy:
	 - Train: 1.0000 
	 - Test: 0.9865
Random Forest F1 Score:
	 - Train: 1.0000
	 - Test: 0.9877


Los resultados muestran que el modelo de Random Forest tiene un rendimiento excelente, con precisión y F1-score prácticamente perfectos en entrenamiento y muy altos en test. La diferencia entre `train` y `test` es mínima, por lo que **no hay sobreajuste**.

Comparando con el árbol de decisión anterior (`accuracy test`: $0.9811$, `F1 test`: $0.9827$), Random Forest obtiene métricas ligeramente mejores. Por tanto, Random Forest es más recomendable para predecir si una persona necesita dieta, aunque ambos modelos son factibles y generalizan bien.

## ENTREGABLE II

1. Se requiere realizar un modelo utilizando las técnicas de MLP y SVM para analizar la variable EJERCICIO. Eliminar del conjunto de datos la variable DIETA para realizar el ejercicio. Sabiendo esto, ¿el modelo será de clasificación o de regresión? Analiza los datos y transformarlos si fuera necesario. Justifica las decisiones que tomes.

2. Realiza una validación cruzada y crea un modelo MLP estableciendo como mínimo 2 parámetros. Utiliza dos métricas ¿Hay sobreajuste o subajuste? Justifica la respuesta.

3. Realiza un hold-out 80-20 y crea una SVM con un kernel de base radial, estableciendo dos parámetros como mínimo. Obten como mínimo 2 métricas. ¿Hay sobreajuste o subajuste? ¿Es más fiable la técnia MLP o la SVM para saber la media de ejercicio de cardio que debe de realizar una persona a la semana? Justifica tus respuestas.

## ENTREGABLE III

1. Se requiere realizar un modelo clústering utilizando las variables Edad, Género, Altura, Peso, Verduras y NComidas. Analiza estos datos y realiza las transformaciones que fueran necesaria. 

2. Aplica la técnica Kmeans justificando y calculando antes cuál es el número de clústeres óptimos mediante la regla del codo ¿Cuál es el número óptimo de clúster? Justifica las respuestas. **Muestra** el número de **clusters** óptimo y la asignación de clusters resultante.

3. Calcula el número de cluster óptimo utilizando el índice de David Davies-Bouldin. Muestra el número de clusters óptimo y la asignación de clusters resultante ¿Cómo de bueno o malo es el número de clústers óptimo, según el índice utilizado? ¿Cúal de los de las dos clusters generados (preguntas 2 y3) consideras mejor? Justifica tus respuestas.

## ENTREGABLE IV

1. Se requiere realizar un modelo para predecir la variable **Dieta** utilizando Random Forest seleccionando las 5 variables más importantes ¿El modelo será de clasificación o de regresión? Analiza los datos y transformarlos si fuera necesario ¿El modelo con las 5 variables más importantes mejora los resultados? Justifica las decisiones que tomes.

2. Utiliza la función RFECV de sklearn en una validación cruzada de 5 folds para seleccionar las variables más relevantes del modelo. Mediante un hold-out hold-out de 70-30 compara el rendimiento, utilizando como mínimo dos métricas, de  un árbol de decisión con todas las características y un árbol de decisión sólo con las seleccionadas ¿Qué modelo es mejor? ¿Es factible el modelo para saber si una persona tiene que hacer dieta o no?